### PR TITLE
ManagedVec sort & deduplicate

### PR DESCRIPTION
**Pull request by alyn509**

### CODE REVIEW

File: framework/base/src/types/managed/wrapped/cached_managed_buffer.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+fn load_static_cache<M>(managed_buffer: &ManagedBuffer<M>) -> StaticBufferRef<M>
+where
+    M: ManagedTypeApi,
+{
+    StaticBufferRef::try_new_from_copy_bytes(managed_buffer.len(), |dest_slice| {
+        let _ = managed_buffer.load_slice(0, dest_slice);
+    })
+    .unwrap_or_else(|| M::error_api_impl().signal_error(b"Static cache is in use"))
```
Comment (claudiu725) [inline framework/base/src/types/managed/wrapped/cached_managed_buffer.rs:13]: Slightly misleading error message - this error can also occur when the static buffer is too small for loading the managed buffer (the limit is 10000 bytes at the moment) - see [try_lock_with_copy_bytes](https://github.com/ElrondNetwork/elrond-wasm-rs/blob/7cf42c8e03c8b0543a4ca51ebf028b5416166d9d/framework/base/src/types/static_buffer/lockable_static_buffer.rs#L39)

**Review by claudiu725 — CHANGES_REQUESTED** [review]

File: framework/base/src/types/managed/wrapped/cached_managed_buffer.rs (hunk 1 above)
Comment (claudiu725) [reply framework/base/src/types/managed/wrapped/cached_managed_buffer.rs:13]:
```suggestion
    .unwrap_or_else(|| M::error_api_impl().signal_error(b"static cache too small or already in use"))
```

**Review by claudiu725 — APPROVED** [review]

**Review by dorin-iancu — DISMISSED** [review]

File: framework/base/src/types/managed/wrapped/managed_vec.rs [hunk 2]
```diff
@@ -291,6 +296,165 @@ where
...
+            Self::reinterpret_slice(result, result_len)
+        });
+    }
+
+    fn reinterpret_slice<T1, T2>(from: &[T1], len: usize) -> &[T2] {
+        let ptr = from.as_ptr() as *const T2;
+        unsafe { core::slice::from_raw_parts(ptr, len) }
+    }
+
+    fn reinterpret_slice_mut<T1, T2>(from: &mut [T1], len: usize) -> &mut [T2] {
```
Comment (andrei-marinica) [inline framework/base/src/types/managed/wrapped/managed_vec.rs:336]: Also unsafe

File: framework/base/src/types/managed/wrapped/cached_managed_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+{
+    StaticBufferRef::try_new_from_copy_bytes(managed_buffer.len(), |dest_slice| {
+        let _ = managed_buffer.load_slice(0, dest_slice);
+    })
+    .unwrap_or_else(|| {
+        M::error_api_impl().signal_error(b"static cache too small or already in use")
+    })
+}
+
+pub fn with_buffer_contents<M, R, F>(managed_buffer: &ManagedBuffer<M>, f: F) -> R
```
Comment (andrei-marinica) [inline framework/base/src/types/managed/wrapped/cached_managed_buffer.rs:18]: I would make this a method of ManagedBuffer. Easier to find.

File: framework/base/src/types/managed/wrapped/cached_managed_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+use crate::{
+    api::{ErrorApiImpl, ManagedTypeApi},
+    types::{ManagedBuffer, StaticBufferRef},
+};
+
+fn load_static_cache<M>(managed_buffer: &ManagedBuffer<M>) -> StaticBufferRef<M>
```
Comment (andrei-marinica) [inline framework/base/src/types/managed/wrapped/cached_managed_buffer.rs:6]: I would make this a method of ManagedBuffer. Easier to find.

**Review by andrei-marinica — CHANGES_REQUESTED** [review]

**Review by andrei-marinica — DISMISSED** [review]

**Review by claudiu725 — APPROVED** [review]

**Pull request merged**